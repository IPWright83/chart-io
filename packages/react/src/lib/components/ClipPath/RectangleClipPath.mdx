import { Story, Canvas, Meta } from "@storybook/blocks";

<Meta title="Components/RectangleClipPath" />

# RectangleClipPath Component

The `<RectangleClipPath>` component allows us to create a clip path to ensure that any artifacts that appear outside the visible range are not rendered. This is especially useful when zooming into the domain on a chart so is frequently used with Brushes.

<Canvas>
    <Story id="components-rectangleclippath--default" />
</Canvas>

## Adding a RectangleClipPath

To use a `<RectangleClipPath>` component you should wrap it around all your children that require clipping. For example you may wrap several plots like so:

```
<RectangleClipPath>
    <Line x="x" y="y" />
    <Line x="x" y="y2" />
</RectangleClipPath>
```

This will dynamically inject a `clipPath` prop into the components which can be consumed using the form:

```
const MyClippedComponent = ({ clipPath }: { clipPath?: string }) => {
    return <circle clipPath={`url(#${clipPath})`} ... />;
};
```
