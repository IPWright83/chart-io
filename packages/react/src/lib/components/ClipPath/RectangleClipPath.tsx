import React from "react";

import { useSelector } from "react-redux";

import { chartSelectors, IState } from "../../store";
import { extendChildrenProps } from "../../utils";

export interface IRectangleClipPathProps {
    /**
     * The child components for the chart
     */
    children?: JSX.Element | JSX.Element[];
}

/**
 * This component renders a Rectangular ClipPath that can be used to prevent overspill on the chart
 * @return The ClipPath component
 */
export function RectangleClipPath({ children }: IRectangleClipPathProps) {
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const margin = useSelector((s: IState) => chartSelectors.dimensions.margin(s));
    const chartID = useSelector((s: IState) => chartSelectors.id(s));

    const plotWidth = Math.max(0, width - margin.left - margin.right);
    const plotHeight = Math.max(0, height - margin.top - margin.bottom);

    const childrenWithProps = extendChildrenProps(children, { clipPath: `clip-path-${chartID}` });

    return (
        <>
            <defs>
                <clipPath id={`clip-path-${chartID}`}>
                    <rect width={plotWidth} height={plotHeight} x={margin.left} y={margin.top} />
                </clipPath>
            </defs>
            {childrenWithProps}
        </>
    );
}
