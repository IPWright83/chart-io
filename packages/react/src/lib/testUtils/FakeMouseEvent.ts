export interface IFakeMouseEventValues {
    pageX?: number;
    pageY?: number;
    offsetX?: number;
    offsetY?: number;
    x?: number;
    y?: number;
}

/**
 * Create a Fake MouseEvent for testing
 */
export class FakeMouseEvent extends MouseEvent {
    /**
     * Construct a new instance of a FakeMoustEvent
     * @param  type        The type of event, e.g. "click", "mouseenter"
     * @param  values      An object containing values for the x,y co-ordinates for the event
     */
    constructor(type: string, values: IFakeMouseEventValues) {
        const { pageX, pageY, offsetX, offsetY, x, y, ...mouseValues } = values;
        super(type, mouseValues);

        Object.assign(this, {
            offsetX: offsetX || 0,
            offsetY: offsetY || 0,
            pageX: pageX || 0,
            pageY: pageY || 0,
            x: x || 0,
            y: y || 0,
        });
    }
}
