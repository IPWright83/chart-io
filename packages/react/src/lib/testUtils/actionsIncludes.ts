/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
