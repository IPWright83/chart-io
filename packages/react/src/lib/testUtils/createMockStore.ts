import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
