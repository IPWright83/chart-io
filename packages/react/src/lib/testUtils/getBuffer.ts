/**
 * Returns the image buffer for a Canvas
 * @param canvas    The HTML canvas that we're going to extract the buffer from
 * @returns         The image buffer
 */
export function getBuffer(canvas: HTMLCanvasElement) {
    const image = canvas.toDataURL("image/png");
    const data = image.replace(/^data:image\/\w+;base64,/, "");
    // eslint-disable-next-line no-undef
    return Buffer.from(data, "base64");
}
