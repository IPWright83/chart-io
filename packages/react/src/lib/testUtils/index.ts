export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
