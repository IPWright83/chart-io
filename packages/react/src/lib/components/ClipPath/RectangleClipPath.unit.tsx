import React from "react";

import { renderChart, wait } from "../../testUtils";

import { RectangleClipPath } from "./RectangleClipPath";

describe("RectangleClipPath", () => {
    const Circle = ({ clipPath }: { clipPath?: string }) => {
        return <circle clipPath={`url(#${clipPath})`} cx={0} cy={0} r={100} fill="red" />;
    };

    it("renders correctly", async () => {
        const { asFragment } = await renderChart({
            children: (
                <RectangleClipPath>
                    <Circle />
                </RectangleClipPath>
            ),
        });

        await wait(10);

        expect(asFragment()).toMatchSnapshot();
    });
});
