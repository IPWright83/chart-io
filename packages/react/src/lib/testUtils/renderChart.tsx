import type { IData, IScale } from "@chart-io/types";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "./createMockStore";
import { wait } from "./wait";

import { themes } from "../themes";

export async function renderChart({
    children,
    data,
    scales = {},
    store,
}: {
    children: JSX.Element;
    data?: IData;
    scales?: Record<string, IScale>;
    store?: any;
}) {
    const mockStore = createMockStore({
        chart: {
            animationDuration: 0,
            dimensions: {
                height: 200,
                width: 200,
            },
            data,
            scales: Object.entries(scales).reduce((result, [key, scaleFunc]) => {
                result[key] = {
                    domain: scaleFunc.domain(),
                    range: scaleFunc.range(),
                    scale: scaleFunc,
                };

                return result;
            }, {}),
            theme: themes.light,
        },
    });

    const { asFragment, container } = render(
        <Provider store={store ?? mockStore}>
            <svg>{children}</svg>
        </Provider>
    );

    await wait(10);

    return { asFragment, container, store: store ?? mockStore };
}
