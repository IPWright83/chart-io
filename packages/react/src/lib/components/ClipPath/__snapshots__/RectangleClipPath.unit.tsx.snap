// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RectangleClipPath renders correctly 1`] = `
<DocumentFragment>
  <svg>
    <defs>
      <clippath
        id="clip-path-test"
      >
        <rect
          height="400"
          width="800"
          x="0"
          y="0"
        />
      </clippath>
    </defs>
    <circle
      clip-path="url(#clip-path-test)"
      cx="0"
      cy="0"
      fill="red"
      r="100"
    />
  </svg>
</DocumentFragment>
`;
